libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -1759,6 +1782,101 @@ IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
+                                        const IotMqttPublishInfo_t * pPublishInfo )
+{
+    IotMqttError_t status = IOT_MQTT_SUCCESS;
+    uint8_t ** pPacketIdentifierHigh = NULL;
+    uint8_t * pMqttPacket = NULL;
+    size_t packetSize = 0;
+    uint16_t packetIdentifier;
+
+    IotMqtt_Assert( pPublishInfo != NULL );
+
+    /* Check whether the connection handle is valid. */
+    if( mqttConnection->disconnected == true )
+    {
+        status = IOT_MQTT_NETWORK_ERROR;
+    }
+
+    if( status == IOT_MQTT_SUCCESS )
+    {
+        /* Generate a PUBLISH packet from pPublishInfo. */
+        status = _getMqttPublishSerializer( mqttConnection->pSerializer )( pPublishInfo,
+                                                                           &( pMqttPacket ),
+                                                                           &( packetSize ),
+                                                                           &( packetIdentifier ),
+                                                                           pPacketIdentifierHigh );
+    }
+
+    if( status == IOT_MQTT_SUCCESS )
+    {
+        /* Transmit the MQTT packet from the operation over the network. */
+        if( mqttConnection->pNetworkInterface->send( mqttConnection->pNetworkConnection,
+                                                     pMqttPacket,
+                                                     packetSize ) != packetSize )
+        {
+            status = IOT_MQTT_NETWORK_ERROR;
+        }
+    }
+
+    /* Clean up memory for PUBLISH packet. */
+    if( pMqttPacket != NULL )
+    {
+        IotMqtt_FreeMessage( pMqttPacket );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+IotMqttError_t IotMqtt_PublishAsync( IotMqttConnection_t mqttConnection,
+                                     const IotMqttPublishInfo_t * pPublishInfo,
+                                     uint32_t flags,
+                                     const IotMqttCallbackInfo_t * pCallbackInfo,
+                                     IotMqttOperation_t * const pPublishOperation )
+{
+    IotMqttError_t status = IOT_MQTT_SUCCESS;
+
+    /* Check that IotMqtt_Init was called. */
+    if( _checkInit() == false )
+    {
+        status = IOT_MQTT_NOT_INITIALIZED;
+    }
+    /* Check whether the connection handle is valid. */
+    else if( mqttConnection->disconnected == true )
+    {
+        status = IOT_MQTT_NETWORK_ERROR;
+    }
+    else if( _IotMqtt_ValidatePublish( mqttConnection->awsIotMqttMode,
+                                       pPublishInfo,
+                                       flags,
+                                       pCallbackInfo,
+                                       pPublishOperation ) == false )
+    {
+        status = IOT_MQTT_BAD_PARAMETER;
+    }
+    else if( ( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
+               == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) &&
```

Comment:
Are we expecting that the flag MQTT_INTERNAL_FLAG_BLOCK_ON_SEND  is set from application? Isn't this an internal flag and I don't see it in the public docs as well https://docs.aws.amazon.com/freertos/latest/lib-ref/c-sdk/mqtt/mqtt_constants.html#mqtt_constants_flags

Reply:
Good catch. Will remove this logic, was added in the initial revision of the fix in the PR, but had forgotten to remove it.

Regarding your point that the flag is internal facing, unfortunately, the `IotMqtt_PublishAsync` API will accept the flag even if the user accidentally sets the bit associated with this flag in the `flags` input parameter. (This is the behavior of `v4_beta` branch)

Reply:
I see that it is a possible behavior for the async API. Isn't this flag supposed to be set internally though for a publish sync API call for QoS0? Are you missing that set logic in this PR?

Reply:
As discussed, I see that the flag is set from publish sync API